# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [31]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [32]:
os.getcwd()

'C:\\Shalin\\Springboard bootcamp\\projects\\Guided Capstone project 1'

In [33]:
os.listdir()

['.ipynb_checkpoints',
 'data',
 'figures',
 'GuidedCapstoneStep3HL.ipynb',
 'GuidedCapstoneStep4 and step5HL.ipynb',
 'GuidedCapstoneStep6HL.ipynb',
 'GuidedCapstone_Step2HL.ipynb',
 'models',
 'Step 1 -Problem Identification',
 'Step 2 - Data Wrangling',
 'Step 3 - EDA',
 'Step 4 - Data preprocessing and training data development',
 'Step 5 - Modeling',
 'Step 6 - Documentation',
 'updated_ski_data.csv']

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [34]:
filename = 'C:\\Shalin\\Springboard bootcamp\\projects\\Guided Capstone project 1\\data\\step3_output.csv'
df = pd.read_csv(filename)

df.head(5)

Name       state  summit_elev  vertical_drop  \
0            Eaglecrest Ski Area      Alaska         2600           1540   
1               Hilltop Ski Area      Alaska         2090            294   
2            Sunrise Park Resort     Arizona        11100           1800   
3  Yosemite Ski & Snowboard Area  California         7800            600   
4                  Bear Mountain  California         8805           1665   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       0  ...   
1      0        0.0          0          0     0       1  ...   
2      0        0.0          0          1     2       3  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              640.0            60.0              45.0       44.0   
1               30.0            30.0             150.0       36.0   
2              800.0            80.0             115.0       49.0   
3               88.0             0.0             110.0       84.0   
4              198.0           198.0             122.0       76.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            350.0     47.000000      53.00000               90.0   
1             69.0     30.000000      34.00000              152.0   
2            250.0     74.000000      78.00000              104.0   
3            300.0     47.000000      47.00000              107.0   
4            100.0     57.916957      64.16681              130.0   

   NightSkiing_ac  clusters  
0             0.0         0  
1            30.0         0  
2            80.0         1  
3             0.0         2  
4             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [35]:
df.dtypes

Name                  object
state                 object
summit_elev            int64
vertical_drop          int64
trams                  int64
fastEight            float64
fastSixes              int64
fastQuads              int64
quad                   int64
triple                 int64
double                 int64
surface                int64
total_chairs           int64
Runs                 float64
TerrainParks         float64
LongestRun_mi        float64
SkiableTerrain_ac    float64
Snow Making_ac       float64
daysOpenLastYear     float64
yearsOpen            float64
averageSnowfall      float64
AdultWeekday         float64
AdultWeekend         float64
projectedDaysOpen    float64
NightSkiing_ac       float64
clusters               int64
dtype: object

In [36]:
dfo=df.select_dtypes(include=['object']) # select object type columns
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

In [37]:
df.columns

Index(['summit_elev', 'vertical_drop', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface',
       ...
       'state_Rhode Island', 'state_South Dakota', 'state_Tennessee',
       'state_Utah', 'state_Vermont', 'state_Virginia', 'state_Washington',
       'state_West Virginia', 'state_Wisconsin', 'state_Wyoming'],
      dtype='object', length=329)

In [38]:
df.shape

(270, 329)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [39]:
df = df.loc[:,~df.columns.str.startswith('Name')]
df.columns

Index(['summit_elev', 'vertical_drop', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface', 'total_chairs',
       'Runs', 'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac',
       'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen', 'averageSnowfall',
       'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac',
       'clusters', 'state_Alaska', 'state_Arizona', 'state_California',
       'state_Colorado', 'state_Connecticut', 'state_Idaho', 'state_Illinois',
       'state_Indiana', 'state_Iowa', 'state_Maine', 'state_Maryland',
       'state_Massachusetts', 'state_Michigan', 'state_Minnesota',
       'state_Missouri', 'state_Montana', 'state_Nevada',
       'state_New Hampshire', 'state_New Jersey', 'state_New Mexico',
       'state_New York', 'state_North Carolina', 'state_Ohio', 'state_Oregon',
       'state_Pennsylvania', 'state_Rhode Island', 'state_South Dakota',
       'state_Tennessee', 'state_Utah', 'state_Vermont', '

In [40]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['AdultWeekend'], axis = 1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

C:\Users\gosal\Anaconda3_2\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\gosal\Anaconda3_2\lib\site-packages\ipykernel_launcher.py:14: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  


In [41]:
X.shape

(270, 58)

In [42]:
y.shape

(270,)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [43]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [44]:
X_train.shape

(202, 58)

In [45]:
X_test.shape

(68, 58)

In [46]:
y_train.shape

(202,)

In [47]:
y_test.shape

(68,)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [48]:
#Linear Regresion model
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [54]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [55]:
y_pred.mean()

61.48727815236312

## Review Model Outcomes — Iterate over additional models as needed

In [56]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
evs = explained_variance_score(y_test, y_pred)
evs

0.7367255858069923

In [57]:
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_test, y_pred)
mae

6.667547731169333

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [58]:
model.intercept_

14.861283752792438

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [59]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by = ['Coefficient'], ascending = False)

Coefficient
state_Tennessee       1.882277e+01
state_North Carolina  1.521682e+01
state_New Mexico      1.252487e+01
state_Montana         1.167195e+01
state_Massachusetts   1.021386e+01
state_Maryland        9.431548e+00
state_Arizona         9.089298e+00
state_Nevada          8.779054e+00
state_Virginia        8.549212e+00
state_Ohio            8.197452e+00
state_Idaho           8.066485e+00
state_Rhode Island    7.829260e+00
state_New Jersey      7.211811e+00
state_Colorado        7.036637e+00
state_South Dakota    6.797341e+00
state_Connecticut     5.905158e+00
state_Maine           5.674133e+00
state_Indiana         5.324482e+00
state_Minnesota       5.080584e+00
state_Pennsylvania    5.060416e+00
state_Oregon          4.716017e+00
state_West Virginia   4.209765e+00
state_Alaska          4.199715e+00
state_Illinois        4.150238e+00
state_Wyoming         3.998830e+00
state_New Hampshire   3.452679e+00
state_California      2.838173e+00
state_Iowa            2.669371e+00
state_Utah            2.645268e+00
state_Michigan        1.959789e+00
clusters              1.924751e+00
state_Missouri        1.685099e+00
state_Vermont         1.498877e+00
state_Washington      1.297573e+00
AdultWeekday          6.973865e-01
surface               6.960798e-01
quad                  6.679577e-01
state_Wisconsin       6.106197e-01
triple                5.978863e-01
TerrainParks          5.180097e-01
total_chairs          3.987214e-01
state_New York        3.683820e-01
LongestRun_mi         3.232445e-01
fastQuads             1.047226e-01
Runs                  7.979604e-02
double                6.632028e-02
yearsOpen             4.144968e-02
daysOpenLastYear      1.551197e-02
projectedDaysOpen     1.203200e-02
Snow Making_ac        7.877843e-03
NightSkiing_ac        6.489663e-03
SkiableTerrain_ac     3.168536e-03
vertical_drop         2.953801e-03
summit_elev           6.941656e-04
averageSnowfall       1.910383e-04
fastEight             1.470558e-11
trams                 1.158698e-11
fastSixes             6.525891e-13

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [60]:
# Rebuid model without the 'state' columns
X_train_2= X_train.loc[:,~X_train.columns.str.startswith('state')]
X_test_2= X_test.loc[:,~X_test.columns.str.startswith('state')]
X_train_2.shape

(202, 23)

In [61]:
X_test_2.shape

(68, 23)

In [62]:
y_train.shape

(202,)

In [63]:
y_test.shape

(68,)

In [64]:
#Apply Linear Regresion model on the updated dataframe without state columns
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model_2 = lm.fit(X_train_2,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [65]:
# Make a variable called y_pred_2 and assign it the result of calling predict() on our model variable with parameter X_test_2
y_pred_2 = lm.predict(X_test_2)
y_pred_2

array([ 93.03581079,  81.6768078 ,  43.20081347,  38.66125619,
        56.20004579,  46.72161959,  72.24454629, 107.1517422 ,
        59.36383068,  49.78297577,  67.31212032,  75.95798632,
        44.30818999,  73.10188805,  56.24942764,  57.81538451,
        59.70356497,  56.27030526,  51.80825747,  77.01804426,
        38.47648064,  61.51064714,  46.75196896,  56.288816  ,
        54.36331561,  78.30889861,  90.26668035,  54.26942107,
        52.09940242,  55.6180931 ,  62.51343739,  43.7722026 ,
        59.55666238,  60.92939484,  52.84090187,  81.9049391 ,
        44.84809082,  28.93022776,  72.45739036,  46.20674573,
        62.87068072,  63.49185123,  44.21483085,  75.88252926,
        68.7553718 ,  74.58271908,  87.70115219,  50.69718125,
        82.08501166,  48.62659301,  46.3991855 ,  61.34046781,
        39.02240248,  50.80633453,  90.03927303,  60.84888209,
        42.20201808,  67.1359272 ,  91.22542859,  76.0441241 ,
        36.42393865,  56.07107099,  52.53001732,  64.92

## Review Model Outcomes — Iterate over additional models as needed

In [66]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
evs_2 = explained_variance_score(y_test, y_pred_2)
evs_2

0.7596580348808419

In [67]:
from sklearn.metrics import mean_absolute_error
mae_2 = mean_absolute_error(y_test, y_pred_2)
mae_2

6.613244205230069

Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [68]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X_train_2.columns, columns=['Coefficient']).sort_values(by = ['Coefficient'], ascending = False)

Coefficient
clusters           2.925300e+00
quad               8.944225e-01
AdultWeekday       7.216113e-01
surface            6.595900e-01
TerrainParks       4.798342e-01
total_chairs       4.759169e-01
triple             3.588631e-01
double             2.447261e-01
fastQuads          1.269474e-01
Runs               4.230506e-02
LongestRun_mi      3.313352e-02
yearsOpen          2.941888e-02
daysOpenLastYear   2.126885e-02
projectedDaysOpen  8.708788e-03
NightSkiing_ac     8.111249e-03
Snow Making_ac     4.745888e-03
vertical_drop      4.600389e-03
SkiableTerrain_ac  4.246379e-03
averageSnowfall    2.869571e-03
summit_elev        4.702551e-05
trams              4.565792e-15
fastEight          3.046174e-15
fastSixes          1.110223e-15

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [71]:
# Rebuid model without the 'summit_elev' and 'base_elev' columns
X_train_3 = X_train_2.drop(columns = ['summit_elev'], axis = 1)
#X_train_3= X_train_2.loc[:,~X_train_2.columns.str.startswith('summit_')]
X_test_3 = X_test_2.drop(columns = ['summit_elev'], axis = 1)
#X_train_3.columns

In [72]:
X_train_3.shape

(202, 22)

In [73]:
X_test_3.shape

(68, 22)

In [74]:
#Apply Linear Regresion model on the updated dataframe without state and summit_elev columns
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model_3 = lm.fit(X_train_3,y_train)

In [75]:
# Make a variable called y_pred_3 and assign it the result of calling predict() on our model variable with parameter X_test_3
y_pred_3 = lm.predict(X_test_3)
y_pred_3

array([ 93.08635581,  81.61118183,  43.2448648 ,  38.64372189,
        56.37976554,  46.82155572,  72.33385807, 106.77733025,
        59.51045661,  49.78637248,  67.19469494,  75.98778   ,
        44.40509341,  73.305256  ,  56.43044729,  57.99179097,
        59.77030067,  56.24377213,  51.82348566,  76.95758434,
        38.48247528,  61.49570346,  46.74222179,  56.34044943,
        54.31435896,  78.3099563 ,  90.20394279,  54.21666904,
        52.08600595,  55.5835149 ,  62.47294138,  43.73599157,
        59.55408438,  60.90860372,  52.81144059,  81.86450152,
        44.85375917,  28.93344415,  72.65851271,  46.19147134,
        62.83485916,  63.50054401,  44.22948991,  76.06489619,
        68.73987235,  74.55898902,  87.58423156,  50.64339507,
        82.29025576,  48.58901454,  46.35828182,  61.33120715,
        39.00916257,  50.73884408,  90.09416423,  60.79765503,
        42.22614939,  67.07634376,  91.39688212,  76.04391152,
        36.41018213,  56.12648074,  52.52135964,  64.91

## Review Model Outcomes — Iterate over additional models as needed

In [76]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score
evs_3 = explained_variance_score(y_test, y_pred_3)
evs_3

0.7587871534763161

In [77]:
from sklearn.metrics import mean_absolute_error
mae_3 = mean_absolute_error(y_test, y_pred_3)
mae_3

6.629743096594611

Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [78]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X_train_3.columns, columns=['Coefficient']).sort_values(by = ['Coefficient'], ascending = False)

Coefficient
clusters           3.021642e+00
quad               8.989091e-01
AdultWeekday       7.208570e-01
surface            6.587074e-01
TerrainParks       4.823240e-01
total_chairs       4.807072e-01
triple             3.523301e-01
double             2.456042e-01
fastQuads          1.337796e-01
Runs               4.163455e-02
LongestRun_mi      3.706933e-02
yearsOpen          2.948717e-02
daysOpenLastYear   2.101642e-02
projectedDaysOpen  8.538269e-03
NightSkiing_ac     8.305197e-03
Snow Making_ac     4.618653e-03
vertical_drop      4.556349e-03
SkiableTerrain_ac  4.271081e-03
averageSnowfall    2.972488e-03
trams              7.299716e-15
fastSixes          3.830269e-15
fastEight          1.415534e-15

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

Model Selection:

| Model            | Explained Variance| Mean Absolute Error|Features Dropped                 |
| -----------------|-------------------|--------------------|---------------------------------|                    
| Model 1.         | 0.74              | 6.67               | -                               |
| Model 2.         | 0.76              | 6.61               |'state'                          |
| Model 3.         | 0.76              | 6.63               |'state','summit_elev'|

**Conclusion**
Model 2 is the best model since due to highest explained variance and lowest mean absolute error. Considering different characteristics and vairables that apply to each resort, it is important for Big Mountain management to be able to stick to a model that is able to predict with higest confidence what variables/ factors would drive ski resort lift ticket prices since ticket pricing in order to increase the revenue and offset for the increasing CAPEX it incurred due to addition of a new chair lift device. If you note, all 3 models are almost leading to same % confidence and the error is also somewhat in the same range. But going by the metrics as indicated in above table, Model 2 is best